### ISSUE TITLE
Implement `std::error::Error::source()` properly for `[service]::Error` enums.

### ISSUE THREAD

**Issue by abusch** [awslabs/aws-sdk-rust#784]
### Describe the feature

The `Error` enums in the different services implement `std::error::Error` like so: `impl std::error::Error for Error {}`. I believe they should implement `Error::source()` properly by delegating to the contained error struct. This would be particularly valuable for the `Unhandled` variant.

### Use Case

I was trying to decrypt something using `KMS` and kept getting an `Unhandled error` without any other information. I thought something in my request was wrong but it turned out all it was was a missing permission, but it was hard to realise that as the source of the original error was lost (I was using `anyhow` and relied on it to properly display the chain of causes).

### Proposed Solution

I think the `impl std::error::Error for Error{}` line in `error_meta.rs` should be changed to something like:

```rust
impl std::error::Error for Error {
  fn source(&self) -> Option<&(dyn Error + 'static)> {
        match self {
            Self::Variant1(variant1) => variant1.source(),
            ...
        }
    }
```

### Acknowledgements

- [x] I may be able to implement this feature request
- [ ] This feature might incur a breaking change

### Community Note
* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue, please leave a comment

**Comment by abusch** [comment]
Having a quick look at `aws-smithy-rs`, am I correct in thinking that this might be already fixed? https://github.com/awslabs/smithy-rs/blob/main/codegen-client/src/main/kotlin/software/amazon/smithy/rust/codegen/client/smithy/generators/error/OperationErrorGenerator.kt#L218

**Comment by abusch** [comment]
The PR above only exposes the source (opaquely) for the top-level error enum via the standard `Error::source()` method. It doesn't change any of the existing errors, or any of the `Display` implementation, so hopefully that's OK?

Was there any particular issue I should be testing for?

**Comment by Velfi** [comment]
> I know that we changed how we expose source because it was causing error reporters to log the "causal chain" multiple times. I want to make sure we're not undoing that work.

We'll need to send an error to an error reporter and make sure it looks good. That's all.

**Comment by abusch** [comment]
This should be fixed in 0.27.0.

**Comment by abusch** [comment]
Taking the liberty of closing this as this has been implemented.
